tooling/nargo_cli/src/cli/test_cmd.rs

```diff
@@ -77,29 +88,65 @@ pub(crate) fn run(args: TestCommand, config: NargoConfig) -> Result<(), CliError
     };
 
-    // Configure a thread pool with a larger stack size to prevent overflowing stack in large programs.
-    // Default is 2MB.
-    let pool = rayon::ThreadPoolBuilder::new().stack_size(4 * 1024 * 1024).build().unwrap();
-    let test_reports: Vec<Vec<(String, TestStatus)>> = pool.install(|| {
-        workspace
-            .into_iter()
-            .par_bridge()
-            .map(|package| {
-                run_tests::<Bn254BlackBoxSolver>(
-                    &workspace_file_manager,
-                    &parsed_files,
-                    package,
-                    pattern,
-                    args.show_output,
-                    args.oracle_resolver.as_deref(),
-                    Some(workspace.root_dir.clone()),
-                    Some(package.name.to_string()),
-                    &args.compile_options,
-                )
-            })
-            .collect::<Result<_, _>>()
-    })?;
-    let test_report: Vec<(String, TestStatus)> = test_reports.into_iter().flatten().collect();
+    let mut tests = Vec::new();
+    for package in workspace.into_iter() {
+        let package_tests = collect_package_tests::<Bn254BlackBoxSolver>(
+            &workspace_file_manager,
+            &parsed_files,
+            package,
+            pattern,
+            args.show_output,
+            args.oracle_resolver.as_deref(),
+            Some(workspace.root_dir.clone()),
+            package.name.to_string(),
+            &args.compile_options,
+        )?;
+        tests.extend(package_tests);
+    }
+
+    let num_tests = tests.len();
+
+    let num_threads = std::thread::available_parallelism().ok().map(Into::into).unwrap_or(1);
+
+    let (sender, receiver) = mpsc::channel();
+    let iter = Mutex::new(tests.into_iter());
+
+    thread::scope(|scope| {
+        // Start worker threads
+        for _ in 0..num_threads {
+            // Specify a larger-than-default stack size to prevent overflowing stack in large programs.
+            // (the default is 2MB)
+            thread::Builder::new()
+                .stack_size(4 * 1024 * 1024)
+                .spawn_scoped(scope, || loop {
+                    // Get next test to process from the iterator.
+                    let Some(test) = iter.lock().unwrap().next() else {
+                        break;
+                    };
+                    let test_status = (test.runner)();
+
+                    // It's fine to ignore the result of sending. If the
+                    // receiver has hung up, everything will wind down soon
+                    // anyway.
+                    let _ = sender.send((test.name, test.package_name, test_status));
+                })
+                .unwrap();
+        }
+    });
+
+    let mut test_reports_per_package = HashMap::<String, Vec<(String, TestStatus)>>::new();
+
+    // Print results of tests that already dinished
+    for (test_name, package_name, test_status) in receiver.iter().take(num_tests) {
```

Comment:
I copied all of this code from libtest-mimic 😊 (which is what the issue suggested)

https://github.com/LukasKalbertodt/libtest-mimic/blob/63e4ec13f15c4adc5b36e3cc0c5f49e19b50f480/src/lib.rs#L492-L551

Reply:
To answer the question, I'm not sure the senders will go out of the scope because they would be dropped after this loop.

Reply:
Maybe it would be more obviously useful in a setup like this?
```
let (test_tx, test_rx) = crossbeam::channel::unbounded();
let (result_tx, result_rx) = mpsc::channel::unbounded();

for i in 0 .. num_workers { 
  let mut test_rx = test_rx.clone();
  let mut result_tx = result_tx.clone();
  thread::spawn(|| {
    while Ok(test) = test_rx.recv() {
      let result = todo();
      let _ = result_tx.send(result);
    }
  });
}

// Sequential between packages
for package in collect_packages() {
  let package_tests = collect_tests(package);
  let num_tests = package_tests.len();  
  // Parallel inside packages
  for test in package_tests {
    test_tx.send(test);
  }
  // Collect results
  println!("running {num_tests} tests in {package}");
  for _ in 0 .. num_tests {
    if let Ok(result) = result_rx.recv() {
      println!(result);
    }
  }
  println!("tests completed for {package}");
}
```
That is, for each package we know how many results are coming in, so we can display an interim report, before resuming with the next cohort. And it displays results in a streaming fashion as well, as they come. 

The only danger is if a worker crashes while handling a report, we might never get the expected number of results; but other workers are still online, blocked on waiting for new tasks, and the whole thing is deadlocked.

Reply:
> I'm not sure the sender will go out of the scope because it would be dropped after this loop.

I was actually surprised that you did not have to make a clone of the sender, one for each thread. If you do then the receiver should work as long as there are items in the channel, or if it's empty but some sender instances are still around.

Reply:
> That is, for each package we know how many results are coming in, so we can display an interim report, before resuming with the next cohort. And it displays results in a streaming fashion as well, as they come.

What's the advantage of doing this instead of what there is right now? I wonder if this alternative approach could show the second package output before the first one, if there's a `println` inside Noir, which currently is done outside of this sending/receiving (I'm thinking about improving this but I'm not sure it's worth it).

Reply:
Or, put another way, maybe it could be done in a separate PR (I'm thinking about making `--show-output` send the output of Noir's `println` to a buffer, then display that buffer at the end... but it requires a larger refactor)

Reply:
I was also just an idea of why `.take(num_tests)` would be useful in that we can keep our workers alive while intermittently print a summary of tests, because we know how many they have to complete before we give them more tasks 🤷

Reply:
I stopped pushing the remaining code changed I had in mind. If you want, you can try this alternative approach by pushing new commits here, or making a PR on top of this one, mainly because I think you have a clear idea of how to do it and me trying to do it might not result in exactly what you are thinking. Then we can compare the time to run tests to see if it ends up being faster (or, regardless of this, we could keep this alternative approach if it's better in some way).

Reply:
Now that we [confirmed](https://github.com/noir-lang/noir/pull/6728#discussion_r1876110088) that the report is correct I don't have any problem with displaying the results on the go. Maybe it's slower, maybe not, we can try after this PR if we want, please do finish your what you had in mind 🙏

Reply:
So the only real problem with https://github.com/noir-lang/noir/issues/4489 was that [this lock](https://github.com/noir-lang/noir/blob/c58d69141b54a918cd1675400c00bfd48720f896/tooling/nargo_cli/src/cli/test_cmd.rs#L152-L156) had to happen _inside_ the loop, not before it.